src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -38,67 +39,139 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
-            var builderTemp = _factory.Local(builderTempSymbol);
+            BoundLocal builderTemp = _factory.Local(builderTempSymbol);
 
             // PROTOTYPE(interp-string): Support dynamic creation
-            // var builder = new BuilderType(baseStringLength, numFormatHoles);
-            Debug.Assert(data.Construction is BoundObjectCreationExpression);
-            var builderConstruction = _factory.AssignmentExpression(builderTemp, (BoundExpression)VisitObjectCreationExpression((BoundObjectCreationExpression)data.Construction));
+            // var handler = new HandlerType(baseStringLength, numFormatHoles, ...InterpolatedStringHandlerArgumentAttribute parameters, <optional> out bool handlerIsValid);
+            var construction = (BoundObjectCreationExpression)data.Construction;
+
+            BoundLocal? appendShouldProceedLocal = null;
+            if (data.HasTrailingHandlerValidityParameter)
+            {
+                Debug.Assert(construction.ArgumentRefKindsOpt[^1] == RefKind.Out);
+
+                BoundInterpolatedStringArgumentPlaceholder trailingParameter = data.ArgumentPlaceholders[^1];
+                TypeSymbol localType = trailingParameter.Type!;
+                Debug.Assert(localType.SpecialType == SpecialType.System_Boolean);
+                var outLocal = _factory.SynthesizedLocal(localType);
+                appendShouldProceedLocal = _factory.Local(outLocal);
+
+                AddPlaceholderReplacement(trailingParameter, appendShouldProceedLocal);
+            }
+
+            var handlerConstructionAssignment = _factory.AssignmentExpression(builderTemp, (BoundExpression)VisitObjectCreationExpression(construction));
 
             var usesBoolReturn = data.UsesBoolReturns;
-            var builderPatternExpressions = ArrayBuilder<BoundExpression>.GetInstance(usesBoolReturn ? 2 : node.Parts.Length);
-            builderPatternExpressions.Add(builderConstruction);
-            BoundExpression? appendBinaryExpression = null;
+            var resultExpressions = ArrayBuilder<BoundExpression>.GetInstance(node.Parts.Length + 1);
             foreach (var currentPart in node.Parts)
             {
                 var appendCall = (BoundCall)currentPart;
                 Debug.Assert(usesBoolReturn == (appendCall.Method.ReturnType.SpecialType == SpecialType.System_Boolean));
 
-                var rewrittenArgs = VisitList(appendCall.Arguments);
-                var rewrittenAppendCall = MakeCall(
+                // The append call itself could have an interpolated string conversion that uses the builder local as the receiver
+                // passed to a nested construction call. This needs to affect the current call to ensure side effects are
+                // preserved, but shouldn't affect any subsequent calls.
+                BoundExpression? appendReceiver = builderTemp;
```

Comment:
>builderTemp [](http://example.com/codeflow?start=50&length=11)

It is still not clear to me what mutation we are trying to preserve here. Could you give a concrete example? If we a trying to prevent mutation of the value stored in ```builderTemp```, perhaps we should allocate a copy temp instead?

Reply:
Consider the case of an `AppendFormatted` method that itself has an `InterpolatedStringHandlerArgument("")` attribute on it. The call to `VisitArguments` will replace the receiver with a sequence of saving off the temp and then evaluating the temp again. We don't want the rest of the `AppendFormatted` calls in _this_ interpolated builder to use this new receiver sequence, we want them to continue using the same `BoundLocal` we created above.

Reply:
I'm going to add a test to demonstrate this scenario. Look for `InterpolatedStringHandlerArgumentAttribute_AttributeOnAppendFormatCall` in iteration 4 when I push it.